share/oaipmh/repository.py

```diff
@@ -0,0 +1,175 @@
+        else:
+            share_id = IDObfuscator.encode(work)
+        return 'oai{delim}{repository}{delim}{id}'.format(id=share_id, repository=self.REPOSITORY_IDENTIFIER, delim=self.IDENTIFER_DELIMITER)
+
+    def _do_identify(self, kwargs, renderer):
+        earliest = AbstractCreativeWork.objects.order_by('date_modified').values_list('date_modified', flat=True)
+        return renderer.identify(earliest[0] if earliest.exists() else None)
+
+    def _do_listmetadataformats(self, kwargs, renderer):
+        if 'identifier' in kwargs:
+            self.resolve_oai_identifier(kwargs['identifier'])
+        return renderer.listMetadataFormats(self.FORMATS)
+
+    def _do_listsets(self, kwargs, renderer):
+        if 'resumptionToken' in kwargs:
+            self.errors.append(oai_errors.BadResumptionToken(kwargs['resumptionToken']))
+            return
+        return renderer.listSets(Source.objects.values_list('name', 'long_title'))
+
+    def _do_listidentifiers(self, kwargs, renderer):
+        works, next_token, _ = self._load_page(kwargs)
+        if self.errors:
+            return
+        return renderer.listIdentifiers(works, next_token)
+
+    def _do_listrecords(self, kwargs, renderer):
+        works, next_token, metadataRenderer = self._load_page(kwargs)
+        if self.errors:
+            return
+        return renderer.listRecords(works, next_token, metadataRenderer)
+
+    def _do_getrecord(self, kwargs, renderer):
+        metadataRenderer = self._get_metadata_renderer(kwargs['metadataPrefix'])
+        work = self.resolve_oai_identifier(kwargs['identifier'])
+        if self.errors:
+            return
+        return renderer.getRecord(work, metadataRenderer)
+
+    def _get_metadata_renderer(self, prefix, catch=True):
+        try:
+            return self.FORMATS[prefix](self)
+        except KeyError:
+            if not catch:
+                raise
+            self.errors.append(oai_errors.BadFormat(prefix))
+
+    def _load_page(self, kwargs):
+        if 'resumptionToken' in kwargs:
+            try:
+                queryset, next_token, format, cursor = self._resume(kwargs['resumptionToken'])
+            except (ValueError, KeyError):
+                self.errors.append(oai_errors.BadResumptionToken(kwargs['resumptionToken']))
+        else:
+            queryset = self._record_queryset(kwargs)
+            next_token = self._get_resumption_token(kwargs)
+            metadataRenderer = self._get_metadata_renderer(kwargs['metadataPrefix'])
+            cursor = 0
+        if self.errors:
+            return [], None, None
+        if not queryset.exists():
+            self.errors.append(oai_errors.NoResults())
+            return [], None, None
+        # TODO is there a way to prefetch Sources/Relations/Identifiers just for this slice? https://code.djangoproject.com/ticket/26780
+        works = queryset[cursor:cursor + self.PAGE_SIZE + 1]
+        if len(works) > self.PAGE_SIZE:
```

Comment:
Hmm, good point, didn't realize that. `count()` would be an additional query, though, right? Since the queryset is definitely going to be evaluated at this point, it might be best to evaluate it explicitly with `list()` and check the `len` of the result.